nni/compression/pytorch/quantization_speedup/integrated_tensorrt.py

```diff
@@ -44,346 +38,320 @@ def valid_config(config=None):
-                    tracked_min_output = config[layer.name]['tracked_min_output']
-                    tracked_max_output = config[layer.name]['tracked_max_output']
-                    layer.set_output_type(0, Precision_Dict[a_bits])
-                    out_tensor = layer.get_output(0)
-                    out_tensor.dynamic_range = (tracked_min_output, tracked_max_output)
-
-        # Build engine and do int8 calibration.
-        if trt_version == TRT8:
-            engine = builder.build_engine(network, trt_config)
-        else:
-            engine = builder.build_cuda_engine(network)
-        return engine
+    builder.max_batch_size = 1 # TODO: check whether it is necessary
+
+    trt_config.max_workspace_size = common.GiB(4)
+
+    trt_config.set_flag(trt.BuilderFlag.INT8)
+    trt_config.set_flag(trt.BuilderFlag.FP16)
+    trt_config.set_flag(trt.BuilderFlag.PREFER_PRECISION_CONSTRAINTS)
+
+    # Parse onnx model
+    with open(onnx_model_file, 'rb') as model:
+        if not parser.parse(model.read()):
+            logger.error('ERROR: Fail to parse the ONNX file.')
+            for error in range(parser.num_errors):
+                logger.error(parser.get_error(error))
+            raise ValueError('Failed to parse the ONNX file.')
+
+    config_network_precision(network, config)
+
+    # Build engine and do int8 calibration.
+    engine = builder.build_engine(network, trt_config)
+    return engine
+
+def config_network_to_int8(network):
+    for layer_idx in range(network.num_layers):
+        layer = network.get_layer(layer_idx)
+        layer.precision = trt.int8
+
+def build_engine_with_calib(onnx_model_file, calib, input_shape):
+    """
+    Parameters
+    ----------
+    """
+    builder = trt.Builder(TRT_LOGGER)
+    network = builder.create_network(common.explicit_batch())
+    trt_config = builder.create_builder_config()
+    parser = trt.OnnxParser(network, TRT_LOGGER)
+
+    builder.max_batch_size = input_shape[0]
+    trt_config.max_workspace_size = common.GiB(8)
+    trt_config.set_flag(trt.BuilderFlag.INT8)
+    trt_config.set_flag(trt.BuilderFlag.FP16)
+    trt_config.set_flag(trt.BuilderFlag.PREFER_PRECISION_CONSTRAINTS)
+    trt_config.int8_calibrator = calib
+
+    with open(onnx_model_file, 'rb') as model:
+        if not parser.parse(model.read()):
+            for error in range(parser.num_errors):
+                TRT_LOGGER.log(TRT_LOGGER.ERROR, parser.get_error(error))
+            raise ValueError('Failed to parse the ONNX file.')
+
+    TRT_LOGGER.log(TRT_LOGGER.INFO, f'input number: {network.num_inputs}')
+    TRT_LOGGER.log(TRT_LOGGER.INFO, f'output number: {network.num_outputs}')
+
+    profile = builder.create_optimization_profile()
+    input_name = network.get_input(0).name
+    profile.set_shape(input_name, min=input_shape, opt=input_shape, max=input_shape)
+    trt_config.add_optimization_profile(profile)
+
+    config_network_to_int8(network) # not sure whether it takes effect
```

Comment:
why `not sure`? 🤣

Reply:
because line228 already set int8, I am not sure whether it is necessary to set each layer's precision. It is safe to set it anyway. updated the comment to make it clearer.